File: lang/java/ipc/src/main/java/org/apache/aingle/ipc/generic/GenericResponder.java
/*

 */

package org.apache.aingle.ipc.generic;

import java.io.IOException;

import org.apache.aingle.AIngleRemoteException;
import org.apache.aingle.Protocol;
import org.apache.aingle.Schema;
import org.apache.aingle.generic.GenericData;
import org.apache.aingle.generic.GenericDatumReader;
import org.apache.aingle.generic.GenericDatumWriter;
import org.apache.aingle.io.Decoder;
import org.apache.aingle.io.Encoder;
import org.apache.aingle.io.DatumReader;
import org.apache.aingle.io.DatumWriter;
import org.apache.aingle.ipc.Responder;

/** {@link Responder} implementation for generic Java data. */
public abstract class GenericResponder extends Responder {
  private GenericData data;

  public GenericResponder(Protocol local) {
    this(local, GenericData.get());

  }

  public GenericResponder(Protocol local, GenericData data) {
    super(local);
    this.data = data;
  }

  public GenericData getGenericData() {
    return data;
  }

  protected DatumWriter<Object> getDatumWriter(Schema schema) {
    return new GenericDatumWriter<>(schema, data);
  }

  protected DatumReader<Object> getDatumReader(Schema actual, Schema expected) {
    return new GenericDatumReader<>(actual, expected, data);
  }

  @Override
  public Object readRequest(Schema actual, Schema expected, Decoder in) throws IOException {
    return getDatumReader(actual, expected).read(null, in);
  }

  @Override
  public void writeResponse(Schema schema, Object response, Encoder out) throws IOException {
    getDatumWriter(schema).write(response, out);
  }

  @Override
  public void writeError(Schema schema, Object error, Encoder out) throws IOException {
    if (error instanceof AIngleRemoteException)
      error = ((AIngleRemoteException) error).getValue();
    getDatumWriter(schema).write(error, out);
  }

}

File: lang/csharp/src/apache/ipc/Generic/GenericResponder.cs
/**

 */
using System;
using AIngle.Generic;
using Decoder = AIngle.IO.Decoder;
using Encoder = AIngle.IO.Encoder;

namespace AIngle.ipc.Generic
{
    public abstract class GenericResponder : Responder
    {
        protected GenericResponder(Protocol protocol)
            : base(protocol)
        {
        }

        static protected DatumWriter<Object> GetDatumWriter(Schema schema)
        {
            return new GenericWriter<Object>(schema);
        }

        static protected DatumReader<Object> GetDatumReader(Schema actual, Schema expected)
        {
            return new GenericReader<Object>(actual, expected);
        }

        public override object ReadRequest(Schema actual, Schema expected, Decoder input)
        {
            return GetDatumReader(actual, expected).Read(null, input);
        }

        public override void WriteResponse(Schema schema, object response, Encoder output)
        {
            GetDatumWriter(schema).Write(response, output);
        }
    }
}
